src/modules/compliance/src/lib/procedures/packageInstalled.cpp

```diff
@@ -3,31 +3,357 @@
 
 #include <CommonUtils.h>
 #include <Evaluator.h>
+#include <Result.h>
+#include <cstdio>
+#include <fstream>
+#include <linux/limits.h>
+#include <set>
 #include <string>
+#include <unistd.h>
+
+namespace
+{
+
+using compliance::Error;
+using compliance::Result;
+
+typedef struct
+{
+    time_t lastUpdateTime;
+    std::string packageManager;
+    std::set<std::string> packageNames;
+} PackageCache;
+
+static constexpr long PACKAGELIST_TTL = 3000L;        // just shy of an hours
+static constexpr long PACKAGELIST_STALE_TTL = 12600L; // 3.5 hours
+static constexpr char PACKAGE_CACHE_PATH[] = "/var/lib/osconfig/packageCache";
+
+std::string DetectPackageManager(OsConfigLogHandle log)
+{
+    if (ExecuteCommand(NULL, "dpkg -l dpkg", false, false, 0, 0, NULL, NULL, log) == 0)
+    {
+        return "dpkg";
+    }
+    else if (ExecuteCommand(NULL, "rpm -qa rpm", false, false, 0, 0, NULL, NULL, log) == 0)
+    {
+        return "rpm";
+    }
+    return "";
+}
+
+Result<PackageCache> LoadPackageCache(const std::string& path)
+{
+    PackageCache cache;
+    const std::string pkgCacheHeader = "# PackageCache "; // "# PackageCache <packageManager>@<timestamp>\n"
+    cache.lastUpdateTime = 0;
+    cache.packageManager = "";
+    cache.packageNames.clear();
+    std::ifstream cacheFile(path);
+    if (!cacheFile.is_open())
+    {
+        return Error("Failed to open cache file: " + path);
+    }
+
+    std::string header;
+    if (!std::getline(cacheFile, header) || (0 != header.rfind(pkgCacheHeader, 0)))
+    {
+        return Error("Invalid cache file format");
+    }
+
+    auto separatorPos = header.find('@');
+    if (std::string::npos == separatorPos)
+    {
+        return Error("Invalid cache file header format");
+    }
+
+    cache.packageManager = header.substr(pkgCacheHeader.length(), separatorPos - pkgCacheHeader.length());
+    try
+    {
+        cache.lastUpdateTime = std::stol(header.substr(separatorPos + 1));
+    }
+    catch (const std::exception&)
+    {
+        return Error("Invalid timestamp in cache file header");
+    }
+
+    std::string packageName;
+    while (std::getline(cacheFile, packageName))
+    {
+        if (!packageName.empty())
+        {
+            cache.packageNames.insert(packageName);
+        }
+    }
+
+    if (cacheFile.bad())
```

Comment:
Some places in this file check for `bad()`, others `fail()`, others `operator bool`. Are you able to distinguish which errors are captured by which? If not then I suggest checking `if (!cacheFile)` or `if (cacheFile)` consistently, which captures any error state. My reference for this stuff is the table here: https://en.cppreference.com/w/cpp/io/basic_ios/bad

Reply:
Replaced fail() with !, we still need bad() as we getline sets failbit and we don't want it to fail